### ISSUE TITLE
[Bug]: Inconsistent behavior with `Notifications.requestPermission(true)` when user prompted to change settings

### ISSUE THREAD

**Issue by kpturner** [OneSignal/OneSignal-Cordova-SDK#968]
### What happened?

If I call `Notifications.requestPermission(true)` on iOS or Android 13, and the call deems that it is necessary to prompt the user, two things happen:
1) It prompts for permission or
2) If 1) has already been previously prompted, it will prompt the user to change their settings

If 2) takes place, then the call to the function exits as soon as the user clicks the "Settings" button and returns `false` - so you do not actually know if the user allowed notifications in their settings or not.

On Android 12, the function does not exit when the user clicks "Settings".  In fact it does not return until the user has visited their settings, changed them and then return control to the app. At that point the function exits, but always returns `false` regardless of what the user did. 

So it seems impossible to know if the user has allowed notifications in their settings or not after calling `requestPermission(true)`

With Android 13 and iOS you can at least have a listener on the `permissionChange` event which fires when the user changes the permission in the settings.  The same does not work with Android 12.  The event does fire (seemingly inconsistently) but provides completely the wrong value - always `false` even though permission has been granted. 

### Steps to reproduce?

```Markdown
Described above
```

### What did you expect to happen?

I expect Android 12 and 13 and iO to behave consistently.

Ideally, the `requestPermission(true)` call would only return when the user has changed their settings, and will provide the correct value.  

Also, the `permissionChange` event should fire consistently in Android 12 and provide the correct value.

### OneSignal Cordova SDK version

5.0.4

### Which platform(s) are affected?

- [X] iOS
- [X] Android

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by louis123562** [comment]
I can confirm this behaviour. I am not even able to subscribe anymore on Android, it always unsubscribes when calling ```optIn()``` or ```requestPermission()```

**Comment by kpturner** [comment]
> I can confirm this behaviour. I am not even able to subscribe anymore on Android, it always unsubscribes when calling `optIn()` or `requestPermission()`

When you say it unsubscribes you, do you mean that your subscription disappears from the dashboard or it gets flagged as "opted-out"?  Is this Android 12 or 13 by the way?

**Comment by louis123562** [comment]
It unsubscribes every time, i tested android 13 (sdk 33) only.
It only shows "Permission not granted" when i uninstall the app, install it and call ```requestPermission(true)```, then deny it.

It updates the timestamp of 'Last unsubscribed' every time i call ```requestPermission(true)``` or ```requestPermission()```. Very annoing.
<img width="736" alt="image" src="https://github.com/OneSignal/OneSignal-Cordova-SDK/assets/36068395/81222483-ed48-4edf-b425-aaa7e1d68315">

**Comment by louis123562** [comment]
What login function?
```OneSignal.???.login() ?```

**Comment by kpturner** [comment]
Yes I am just trying to see if I can match those symptoms (which seem even more buggy than the ones I reported) but my sequence of events is:
`OneSignal.initialize()`
`OneSignal.login`
and then, if I somehow managed to find out if the user has granted permission or not
`OneSignal.User.pushSubscription.optIn()`

But if I can get the same result as you then things are even worse than I reported here.

**Comment by kpturner** [comment]
As an aside, you should be aware of this issue too: https://github.com/OneSignal/OneSignal-Cordova-SDK/issues/918
which renders most of the functions (that should be async and are not) to return incorrect values when the app starts (which is often when you most need them):
`OneSignal.User.pushSubscription.optedIn`
`OneSignal.User.pushSubscription.id`
`OneSignal.Notifications.hasPermission` 
and add this in for good measure: https://github.com/OneSignal/OneSignal-Cordova-SDK/issues/967

But that aside, when I do this
```
OneSignal.initialize()
OneSignal.login()
OneSignal.User.pushSubscription.optIn()
```
I do end up with this, but it takes quite a while to update correctly in the dashboard
![image](https://github.com/OneSignal/OneSignal-Cordova-SDK/assets/308646/9d30227f-b2dc-4d77-949b-d4f13696bf68)

**Comment by nan-li** [comment]
Hi @kpturner, thank you for reporting, I was able to reproduce the behavior with calling `requestPermission()` in combination with the Settings prompt.

We will investigate a solution.

**Comment by nan-li** [comment]
Hi @louis123562, can you tell me more about your issue? 

When you prompt for permission **but before a choice is made**, the SDK may believe permission is not granted yet and send an update to the server causing the "Last Unsubscribed" date to be  updated on the dashboard. Then when a choice is made, a subsequent request is sent with that selection.

Are you seeing it does not update in the dashboard after you give permission in the permission prompt?

**Comment by kpturner** [comment]
Just top be clear on the Android 12 situation:  If you are prompted to change your settings, even if you allow them, `hasPermission` will continue to return false until you restart the app. It is an absolute nightmare!

**Comment by louis123562** [comment]
> Hi @louis123562, can you tell me more about your issue?
> 
> When you prompt for permission **but before a choice is made**, the SDK may believe permission is not granted yet and send an update to the server causing the "Last Unsubscribed" date to be updated on the dashboard. (This is related to the issue reported by @kpturner that we will investigate.) Then when a choice is made on the permission prompt, a subsequent request is sent with that selection.
> 
> Are you seeing it does not update in the dashboard after you give permission in the permission prompt?

I just tested on Android 14, it behaves the same as Android 13.

When removing permission manually in Android settings, then going back into the app, call ```requestPermission(true)```, the dialog comes up. If i then look at the OneSignal dashboard, it says "Permission Not Granted" with the current timestamp; so it was updated when the dialog came up.

After accepting to receive notifications in the dialog, the ```Subscriptions Status``` is empty, and the ```Last Subscribed``` is set with the timestamp when touch "accept" in the dialog. Weird!

Calling ```optIn()``` at any time has no effect at all.

**Comment by kpturner** [comment]
@louis123562 it is worth noting that apparently `optIn` does not update the dashboard straight away.  It updates a flag locally straight away, but the requests to the OneSignal dashboard (server) are batched up and sent periodically. I am not sure what triggers them to be sent but it is not immediate.  Obviously this does not explain your scenario though.

**Comment by kpturner** [comment]
> Hi @louis123562, can you tell me more about your issue?
> 
> When you prompt for permission **but before a choice is made**, the SDK may believe permission is not granted yet and send an update to the server causing the "Last Unsubscribed" date to be updated on the dashboard. (This is related to the issue reported by @kpturner that we will investigate.) Then when a choice is made on the permission prompt, a subsequent request is sent with that selection.
> 
> Are you seeing it does not update in the dashboard after you give permission in the permission prompt?

Any update on this @nan-li ?

**Comment by jkasten2** [comment]
@kpturner this issue sounds similar to this one:
https://github.com/OneSignal/OneSignal-Android-SDK/issues/2106

Can you try the latest OneSignal SDK and test again? If you still see the same issue can you enable OneSignal VERBOSE logging and capture the logcat? Mainly looking for the REST API calls make to OneSignal.

**Comment by atwallis** [comment]
I am getting the similar behavior on Android 12 for v5.2.6 of the package. Here is my code. The subscriptions say "Permission denied" and there was never a request for permissions.

```typescript
try {
	console.log('Initializing OneSignal');
	OneSignal.Debug.setLogLevel(LogLevel.Verbose);
	OneSignal.initialize(APP_ID);
	await OneSignal.Notifications.requestPermission(true);
	await OneSignal.login(USER_ID);
} catch (e) {
	console.error(e);
}
OneSignal.Notifications.addEventListener('click', async (e) => {
	const notification = e.notification;
	const url = notification.launchURL;
	console.log({
		notification,
		url
	});
});
console.log("OneSignal initialized");
```

<img width="880" alt="image" src="https://github.com/user-attachments/assets/a010f590-30f8-480a-923a-b58a30ac5a7c">
